### PR TITLE
feat(BACK-1597): add maverickV2 buy support

### PR DESCRIPTION
**Pull request by Wertusser**

### CODE REVIEW

File: src/dex/maverick-v2/maverick-v2.ts [hunk 1]
```diff
@@ -318,7 +333,9 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
     tokenAddress: Address,
     limit: number,
   ): Promise<PoolLiquidity[]> {
-    const _tokenAddress = this.dexHelper.config.wrapETH(tokenAddress);
+    const _tokenAddress = this.dexHelper.config
+      .wrapETH(tokenAddress)
+      .toLowerCase();
 
     const pools = await this._queryPoolsAPI(SUBGRAPH_TIMEOUT);
```
Comment (KanievskyiDanylo) [inline src/dex/maverick-v2/maverick-v2.ts:340]: We need to cache it for some time, no need to query it for all tokens each time

File: src/dex/maverick-v2/maverick-v2.ts [hunk 2]
```diff
@@ -193,6 +185,9 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
             return MAV_V2_BASE_GAS_COST + MAV_V2_TICK_GAS_COST * Number(t);
           });
 
+          // If pool isn't fully initialized, returns null
+          if (unit === 1n) return null;
```
Comment (KanievskyiDanylo) [inline src/dex/maverick-v2/maverick-v2.ts:189]: What if unit is `0` (when error happened)?

Comment (KanievskyiDanylo) [reply src/dex/maverick-v2/maverick-v2.ts:189]: Also, wasn't able to find where we return `1` if pool is not initialized. Could you clarify where it's done?

File: src/dex/maverick-v2/maverick-v2.ts [hunk 3]
```diff
@@ -262,45 +257,65 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
...
+        targetExchange: pool,
+        dexFuncHasRecipient: true,
+        exchangeData,
+        returnAmountPos: extractReturnAmountPosition(
+          this.maverickV2Iface,
+          'swap',
+          'amountOut',
+        ),
+      };
+    } else {
```
Comment (KanievskyiDanylo) [inline src/dex/maverick-v2/maverick-v2.ts:295]: no need in this `else`

File: src/dex/maverick-v2/maverick-v2.ts (hunk 1 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts:342]: fixed

File: src/dex/maverick-v2/maverick-v2.ts (hunk 2 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts]: This fix wont help. There is a problem with pricing calculation and real amount of reserves - MaverickV2 uses ticks for calculating prices, but reserves in ticks are virtual and does not represent real reserves in pool. 

That could leads  to incorrect pricing on BUY side (ex. user swaps 1 USDT and receives 0.00001 USDC, but pricing in this pool will be ~1:1)

File: src/dex/maverick-v2/maverick-v2.ts [hunk 4]
```diff
@@ -262,44 +258,63 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
...
       exchangeData,
-      returnAmountPos:
-        side === SwapSide.SELL
-          ? extractReturnAmountPosition(
-              this.maverickV2Iface,
-              'swap',
-              'amountOut',
-            )
-          : undefined,
+      returnAmountPos: undefined,
```
Comment (JackieJoo) [inline src/dex/maverick-v2/maverick-v2.ts:317]: I think `returnAmountPos` is supported on Ex3 now

File: tests/utils-e2e.ts [hunk 5]
```diff
@@ -62,13 +62,6 @@ import { v4 as uuid } from 'uuid';
 
 export const testingEndpoint = process.env.E2E_TEST_ENDPOINT;
 
-const testContractProjectRootPath = process.env.TEST_CONTRACT_PROJECT_ROOT_PATH;
```
Comment (JackieJoo) [inline tests/utils-e2e.ts:65]: Why do we need changes in this file?

File: tests/utils.ts [hunk 6]
```diff
@@ -76,52 +76,3 @@ export const sleep = (time: number) =>
   new Promise(resolve => {
     setTimeout(resolve, time);
   });
-
-export const generateDeployBytecode = (
```
Comment (JackieJoo) [inline tests/utils.ts:80]: Why do we need changes in this file?

Comment (Wertusser) [reply tests/utils.ts]: revert this change

File: tests/utils-e2e.ts (hunk 5 above)
Comment (Wertusser) [reply tests/utils-e2e.ts]: revert this change

File: src/dex/maverick-v2/maverick-v2.ts (hunk 4 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts]: added `returnAmountPos` from router

File: src/dex/maverick-v2/maverick-v2.ts (hunk 2 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts]: removed this fix

File: src/dex/maverick-v2/maverick-v2.ts [hunk 7]
```diff
@@ -27,9 +27,15 @@ import { MaverickV2EventPool } from './maverick-v2-pool';
 import { SUBGRAPH_TIMEOUT } from '../../constants';
 import { Interface } from '@ethersproject/abi';
 import MaverickV2PoolABI from '../../abi/maverick-v2/MaverickV2Pool.json';
+import MaverickV2RouterABI from '../../abi/maverick-v2/MaverickV2Router.json';
 import ERC20ABI from '../../abi/erc20.json';
 import { extractReturnAmountPosition } from '../../executor/utils';
+
 const EFFICIENCY_FACTOR = 3;
```
Comment (Wertusser) [inline src/dex/maverick-v2/maverick-v2.ts:34]: Added `MIN_POOL_VOLUME_USD` and `MIN_POOL_LIQUDITY_USD` to filter out bad pools

File: src/dex/maverick-v2/maverick-v2.ts [hunk 8]
```diff
@@ -97,41 +104,38 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
     return null;
   }
 
+  async getEventPools(srcToken: Token, destToken: Token) {
+    srcToken = this.dexHelper.config.wrapETH(srcToken);
+    destToken = this.dexHelper.config.wrapETH(destToken);
```
Comment (aburkut) [inline src/dex/maverick-v2/maverick-v2.ts:109]: @Wertusser I think we need to add const before `srcToken` and `destToken`

File: src/dex/maverick-v2/maverick-v2.ts [hunk 9]
```diff
@@ -97,41 +104,38 @@ export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
...
+
+    return Object.values(this.pools).filter((pool: MaverickV2EventPool) => {
+      return (
+        (pool.tokenA.address.toLowerCase() === srcToken.address.toLowerCase() ||
```
Comment (aburkut) [inline src/dex/maverick-v2/maverick-v2.ts:113]: Let's put `pool.tokenA.address.toLowerCase()`, `srcToken.address.toLowerCase()`, `pool.tokenB.address.toLowerCase()`  and `destToken.address.toLowerCase()` to constants and use these constants within this condition.

**Review by aburkut — APPROVED** [review]

File: src/dex/maverick-v2/maverick-v2.ts (hunk 9 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts]: fixed

File: src/dex/maverick-v2/maverick-v2.ts (hunk 8 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts]: fixed

File: src/dex/maverick-v2/maverick-v2.ts (hunk 7 above)
Comment (Wertusser) [reply src/dex/maverick-v2/maverick-v2.ts:34]: We now supports all Maverick's pools so there is no need in this constants anymore

**Review by JackieJoo — APPROVED** [review]

**Pull request merged**